src/mongo/db/query/query_request.cpp

```diff
@@ -622,9 +623,10 @@ Status QueryRequest::validate() const {
     BSONObjIterator projIt(_proj);
     while (projIt.more()) {
         BSONElement projElt = projIt.next();
-        if (isTextScoreMeta(projElt)) {
+        if (isTextScoreMeta(projElt) || isShardNameMeta(projElt)) {
```

Comment:
I'm not sure that this part of the change is needed. If I'm reading the code correctly, this PR does not add the ability to perform a $meta-sort by "shardName", as in {$sort: {foo: {$meta: "shardName"}}}. Furthermore, the query team is interested in relaxing the rule for "textScore" which says that a "textScore" $meta sort must have a corresponding "textScore" $meta projection. See SERVER-43816.

Reply:
I have duplicated the implementation of `textScore` in most places, that is why this condition was added, removing.